tokio-util/src/task/spawn_pinned.rs

```diff
@@ -0,0 +1,247 @@
+    ///             // Rc is !Send + !Sync
+    ///             let local_data = Rc::new("test");
+    ///
+    ///             // This future holds an Rc, so it is !Send
+    ///             async move { local_data.to_string() }
+    ///         })
+    ///         .await
+    ///         .unwrap();
+    ///
+    ///     assert_eq!(output, "test");
+    /// }
+    /// ```
+    pub fn spawn_pinned<F, Fut>(&self, create_task: F) -> JoinHandle<Fut::Output>
+    where
+        F: FnOnce() -> Fut,
+        F: Send + 'static,
+        Fut: Future + 'static,
+        Fut::Output: Send + 'static,
+    {
+        self.pool.spawn_pinned(create_task)
+    }
+}
+
+impl Debug for LocalPoolHandle {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str("LocalPoolHandle")
+    }
+}
+
+struct LocalPool {
+    workers: Vec<LocalWorkerHandle>,
+}
+
+impl LocalPool {
+    /// Spawn a `?Send` future onto a worker
+    fn spawn_pinned<F, Fut>(&self, create_task: F) -> JoinHandle<Fut::Output>
+    where
+        F: FnOnce() -> Fut,
+        F: Send + 'static,
+        Fut: Future + 'static,
+        Fut::Output: Send + 'static,
+    {
+        let (sender, receiver) = tokio::sync::oneshot::channel();
+
+        let worker = self.find_and_incr_least_burdened_worker();
+        let worker_handle_clone = worker.clone();
+
+        // Spawn a future onto the worker's runtime so can immediately return
+        // a join handle.
+        worker.runtime_handle.spawn(async move {
+            // Inside the future we can't run spawn_local yet because we're not
+            // in the context of a LocalSet. We need to send create_task to the
+            // LocalSet task for spawning.
+            let spawn_task = Box::new(move || {
+                // Once we're in the LocalSet context we can call spawn_local,
+                // but first try to create the task. We use catch_unwind here so
+                // we don't kill the worker if create_task panics.
+                let task = match std::panic::catch_unwind(AssertUnwindSafe(create_task)) {
+                    Ok(task) => task,
+                    Err(e) => {
+                        // Creating the task triggered a panic. Send the error
+                        // back so it gets propagated.
+                        let _ = sender.send(Err(e));
+                        return;
+                    }
+                };
+
+                // Task was created successfully, now spawn it.
+                let join_handle = spawn_local(task);
+
+                // Send the join handle back to the spawner.
+                let _ = sender.send(Ok(join_handle));
+            });
+
+            // Send the callback to the LocalSet task
+            if let Err(e) = worker_handle_clone.spawner.send(spawn_task) {
+                // Failed to spawn the task on the worker, so we need to remove
+                // the task from the count.
+                worker_handle_clone.decrement_task_count();
```

Comment:
This seems rather fragile. How about defining a guard struct in the method and doing this in its destructor?